diff -Nurdp nethack-3.4.3/sys/unix/unixmain.c nethack-3.4.3-easywizard/sys/unix/unixmain.c
--- nethack-3.4.3/sys/unix/unixmain.c	2012-12-27 13:44:38.000000000 +0900
+++ nethack-3.4.3-easywizard/sys/unix/unixmain.c	2012-12-27 13:46:38.000000000 +0900
@@ -316,34 +316,9 @@ char *argv[];
 		switch(argv[0][1]){
 		case 'D':
 #ifdef WIZARD
-			{
-			  char *user;
-			  int uid;
-			  struct passwd *pw = (struct passwd *)0;
-
-			  uid = getuid();
-			  user = getlogin();
-			  if (user) {
-			      pw = getpwnam(user);
-			      if (pw && (pw->pw_uid != uid)) pw = 0;
-			  }
-			  if (pw == 0) {
-			      user = nh_getenv("USER");
-			      if (user) {
-				  pw = getpwnam(user);
-				  if (pw && (pw->pw_uid != uid)) pw = 0;
-			      }
-			      if (pw == 0) {
-				  pw = getpwuid(uid);
-			      }
-			  }
-			  if (pw && !strcmp(pw->pw_name,WIZARD)) {
-			      wizard = TRUE;
-			      break;
-			  }
-			}
-			/* otherwise fall thru to discover */
-			wiz_error_flag = TRUE;
+			/* simply make the player wizard */
+			wizard = TRUE;
+			break;
 #endif
 		case 'X':
 			discover = TRUE;
