diff -Nurdp eix-0.33.2-original/work/eix-0.33.2/configure.ac eix-0.33.2-prepare/work/eix-0.33.2/configure.ac
--- eix-0.33.2-original/work/eix-0.33.2/configure.ac	2018-08-28 15:37:52.853150332 +0900
+++ eix-0.33.2-prepare/work/eix-0.33.2/configure.ac	2018-08-28 15:45:49.277246921 +0900
@@ -1445,7 +1445,10 @@ typedef std::pair<int, int> mypair;
 std::vector<mypair> a;
 a.push_back(mypair(4, 4));
 a.emplace_back(5, 5);
-return (a.size() == 2) ? 0 : 1;
+std::vector<bool> b;
+b.push_back(true);
+b.emplace_back(false);
+return (a.size() == 2 && b.size() == 2) ? 0 : 1;
 		]])],
 		[MV_MSG_RESULT([yes])
 		AC_DEFINE([HAVE_EMPLACE_BACK], [1],
@@ -1462,7 +1465,10 @@ typedef std::pair<int, int> mypair;
 std::set<mypair> a;
 a.insert(mypair(4, 4));
 a.emplace(5, 5);
-return (a.size() == 2) ? 0 : 1;
+set::set<bool> b;
+b.insert(true);
+b.emplace(false);
+return (a.size() == 2 && b.size() == 2) ? 0 : 1;
 		]])],
 		[MV_MSG_RESULT([yes])
 		AC_DEFINE([HAVE_EMPLACE], [1],
